server/datastore/mysql/activities.go

```diff
@@ -288,3 +288,61 @@ func (ds *Datastore) ListHostPastActivities(ctx context.Context, hostID uint, op
 
 	return activities, metaData, nil
 }
+
+func (ds *Datastore) CleanupActivitiesAndAssociatedData(ctx context.Context, maxCount int, expiredWindowDays int) error {
+	const selectActivitiesQuery = `
+		SELECT a.id FROM activities a
+		LEFT JOIN host_activities ha ON (a.id=ha.activity_id)
+		WHERE ha.activity_id IS NULL AND a.created_at < DATE_SUB(NOW(), INTERVAL ? DAY)
+		ORDER BY a.id ASC
+		LIMIT ?;`
+	var activityIDs []uint
+	if err := sqlx.SelectContext(ctx, ds.writer(ctx), &activityIDs, selectActivitiesQuery, expiredWindowDays, maxCount); err != nil {
+		return ctxerr.Wrap(ctx, err, "select activities for deletion")
+	}
+	if len(activityIDs) > 0 {
+		deleteActivitiesQuery, args, err := sqlx.In(`DELETE FROM activities WHERE id IN (?);`, activityIDs)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "build activities IN query")
+		}
+		if _, err := ds.writer(ctx).ExecContext(ctx, deleteActivitiesQuery, args...); err != nil {
+			return ctxerr.Wrap(ctx, err, "delete expired activities")
+		}
+	}
+
+	//
+	// `activities` and `queries` are not tied because the activity itself holds
+	// the query SQL so they don't need to be executed on the same transaction.
+	//
+	if err := ds.withTx(ctx, func(tx sqlx.ExtContext) error {
```

Comment:
It doesn't seem like using a transaction is a must here. Do we prefer using transactions to always keep DB data consistent?

Reply:
I am running a tx because I'm deleting related rows that should be deleted together (and do not have foreign key references probably due to performance issues in the past?).